### PR TITLE
Generate swagger json

### PR DESCRIPTION
**Pull request by sapk**
Use https://github.com/go-swagger/go-swagger with go generate to (auto-)build public/swagger.json

Related : #194 go-gitea/go-sdk#53 

I will fill missing API descriptions. For the responses descriptions, tags must be add to sdk repo (so in a other PR).

You can test rendering at http://editor.swagger.io and copy-paste swagger.json. 

<details>
      <summary>Documented API : </summary>

- [X] GET /version (getVersion)
- [x] POST /markdown (renderMarkdown)
- [x] POST /markdown/raw (renderMarkdownRaw)
- [x] GET /users/search (userSearch)
- [x] GET /users/:username (userGet)
- [x] GET /user (userGetCurrent)
- [x] GET /users/:username/tokens (userGetTokens)
- [x] POST /users/:username/tokens (userCreateToken)
- [x] GET /users/:username/repos (userListRepos)
- [x] GET /user/repos (userCurrentListRepos)
- [x] GET /user/keys userCurrentListKeys
- [x] GET /users/:username/keys userListKeys
- [x] GET /user/keys/:id userCurrentGetKey
- [x] POST /user/keys userCurrentPostKey
- [x] DELETE /user/keys/:id userCurrentDeleteKey
- [x] GET /user/gpg_keys userCurrentListGPGKeys
- [x] GET /users/:username/gpg_keys userListGPGKeys
- [x] GET /user/gpg_keys/:id userCurrentGetGPGKey
- [x] POST /user/gpg_keys userCurrentPostGPGKey
- [x] DELETE /user/gpg_keys/:id userCurrentDeleteGPGKey
- [x] GET /user/followers userCurrentListFollowers
- [x] GET /users/:username/followers userListFollowers
- [x] GET /user/following userCurrentListFollowing
- [x] GET /users/:username/following userListFollowing
- [x] GET /user/following/:username userCurrentCheckFollowing
- [x] GET /users/:username/following/:target userCheckFollowing
- [x] PUT /user/following/:username userCurrentPutFollow
- [x] DELETE /user/following/:username userCurrentDeleteFollow
- [x] GET /users/:username/starred userListStarred
- [x] GET /user/starred userCurrentListStarred
- [x] GET /user/starred/:username/:reponame userCurrentCheckStarring
- [x] PUT /user/starred/:username/:reponame userCurrentPutStar
- [x] DELETE /user/starred/:username/:reponame userCurrentDeleteStar
- [x] GET /users/:username/subscriptions userListSubscriptions
- [x] GET /user/subscriptions userCurrentListSubscriptions
- [x] GET /repos/:username/:reponame/subscription userCurrentCheckSubscription
- [x] PUT /repos/:username/:reponame/subscription userCurrentPutSubscription
- [x] DELETE /repos/:username/:reponame/subscription userCurrentDeleteSubscription
- [x] POST /org/:org/repos createOrgRepo
- ..

</details>

### CODE REVIEW

**Comment by lunny** [comment]
Maybe we need an API docs site? Or a link on docs.gitea.io?

**Comment by sapk** [comment]
Once fully implented we could use swagger ui for presentation. There is two choice possible :
- We use a remote one. This require to have CORS enable to be able to execute test request from ui. (example : http://petstore.swagger.io/)
- We use expose a swagger ui inside gitea and we have complete access to api and docs is accessible alongside gitea. 

As global documentation we could start a docker containing swagger ui (allready avalaibe) for try.gitea.io : https://github.com/swagger-api/swagger-ui

**Comment by sapk** [comment]
I think that I will simply copy https://github.com/swagger-api/swagger-ui/tree/master/dist into public/assets. This will provide swagger-ui on all instance.

**Comment by lunny** [comment]
@sapk so add new `make` subcommand is enough?

**Comment by sapk** [comment]
@lunny For generating swagger.json, I added a go generate here :
https://github.com/go-gitea/gitea/pull/1402/files#diff-2620fdbeeb83ce43692534e6c2c39452R5

So **make generate** will update it and place it under public folder.

For swagger-ui, I am thinking of adding a other step to **make assets**

**Comment by sapk** [comment]
In parallel, I will PR to fix little things like go-gitea/go-sdk#50 (it will need a PR in gitea also) to put all data structure under go-sdk and add missing api request. A other PR to go-gitea/go-sdk will be added to add swagger comment to api response and params.

This PR is a long running one to complete ^^. It will maybe need a second pass to adjust comments but the base will be there.

**Comment by lunny** [comment]
Or maybe we can add a link on the bottom or top on Gitea named `API` to the API docs home page.

File: Makefile
Comment (bkcsoft) [inline Makefile]: maybe use `"s;https://...;...;g"` instead for readability?

File: Makefile
Comment (bkcsoft) [inline Makefile]: Just download a tar.gz release instead. We shouldn't depend on latest master being functional

File: routers/api/v1/misc/markdown.go
Comment (bkcsoft) [inline routers/api/v1/misc/markdown.go]: Maybe update the doc links while you're at it :)

Comment (bkcsoft) [reply routers/api/v1/misc/markdown.go]: Or remove them since they now become redundant 😅

File: vendor/vendor.json
Comment (bkcsoft) [inline vendor/vendor.json]: This needs to be fixed before merge :)

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
Actually, would it make sense to have this in `go-gitea/sdk` instead?

File: vendor/vendor.json
Comment (sapk) [reply vendor/vendor.json]: Yes this temporary and will be added in a PR in sdk. All the input and output format (used by api) are in sdk and I have already done one PR to move a struct in gitea to sdk.

File: Makefile
Comment (sapk) [reply Makefile]: I am not a big expert in sed. ^^

**Comment by lunny** [comment]
Is this still WIP?

**Comment by sapk** [comment]
@lunny yes because it is missing some api url and it need a PR in sdk for all format response and request. But you can review tooling around it like Makefile and the only changes that will append next are swagger comment so It will be ready to merge when complete.

File: Makefile
Comment (sapk) [reply Makefile]: I use git tag now to fix the version. (less things to change ^^)

**Comment by lunny** [comment]
@sapk maybe move this to v1.3?

**Comment by sapk** [comment]
@lunny There is two possibility : 
 - Push it in v1.2, everything is working it just need to doc (add comment) to the not yet documented API in other(s) PR.
 - Wait v1.3 to have complete documentation.

**Comment by bkcsoft** [comment]
@lunny @sapk For API-docs I'm actually fine with having partial docs for a while. Better than to have no docs at all 😅

**Comment by bkcsoft** [comment]
This will have my LG-TM once the gogs docs links are gone 🙂

**Comment by jonasfranz** [comment]
The swagger file is not complete. Some parts of the API are missing like the issue management.

**Comment by sapk** [comment]
@JonasFranzDEV That what I say ^^
@bkcsoft ok I will clean this up for fully documented in swagger

**Comment by sapk** [comment]
I rebase/squash all changes into a fully working version. I vendor go-gitea/go-sdk#53 in the wait of it validation so that this PR is working for validation also. I will revert it back as soon that go-gitea/go-sdk/pull/53 is merged.

For testing, just build gitea like always, start it, and go to `http://localhost:3000/swagger`

Fix/Hack that I have implemented : (This is if someone has a better/cleaner idea)
- Swagger-ui doesn't support recursive deps like GPGKey that contain sub-GPGKey that contain .... -> I use a `sed`in go generate to set sub-GPGKey as object.

Improvements to be done in other PR : 
- Add all missing API
- Improve documentation of parameters and responses. (For example : add a regex for armored key params)
- Add params for in url/path params. As of now they are displayed but we need struct to link them to params in code and be able to set them in swagger-ui.

Over all, this PR should not have an big impact on code functionality as it is mostly comments.

**Comment by lunny** [comment]
@sapk I like first chosen. So is it ready to review and merge?

**Comment by sapk** [comment]
@lunny Yes, the only thing that bother me is the little fix/hack for recursive structure not supported by swagger-ui but we need it. (describe in my previous comment)

**Comment by KangoV** [comment]
Which version are you using? We've just switched to using swagger-ui v3.0.7. Has a much nicer look. This is getting ready for OpenAPI v3.0 later in the year: http://swagger.io/new-swagger-ui-and-swagger-editor/.
It's also much easier to integrate.

**Comment by sapk** [comment]
@KangoV I have added swagger-ui v3.0.7 (get via makefile). I used go-swagger to generate the JSON from code comments.

**Comment by lunny** [comment]
Is this ready for review?

**Comment by sapk** [comment]
@lunny yes like I said two days ago ^^
and since go-gitea/go-sdk#53 is merged so nothing else to do.

**Comment by lunny** [comment]
LGTM

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
LGTM

**Comment by bkcsoft** [comment]
make LG-TM work

**Pull request merged**
